File: linux/src/qmk_toolbox/bootloader/halfkay_device.py
from .bootloader_device import BootloaderDevice
from .bootloader_type import BootloaderType


class HalfKayDevice(BootloaderDevice):
    def __init__(self, usb_device):
        super().__init__(usb_device)
        self.bootloader_type = BootloaderType.HALFKAY
        self.name = "HalfKay"
        self.preferred_driver = "usbhid"
    
    async def flash(self, mcu: str, file_path: str) -> int:
        args = ["-mmcu=" + mcu, "-w", file_path, "-v"]
        return await self._run_process("teensy-loader-cli", args)

File: src/QmkToolbox.Core/Bootloader/Impl/HalfKayDevice.cs
using QmkToolbox.Core.Models;
using QmkToolbox.Core.Services;

namespace QmkToolbox.Core.Bootloader.Impl;

/// <summary>HalfKay bootloader device (Teensy/Ergodox EZ, via teensy_loader_cli).</summary>
internal sealed class HalfKayDevice : BootloaderDevice
{
    public HalfKayDevice(IUsbDevice device, IFlashToolProvider toolProvider)
        : base(device, toolProvider)
    {
        Type = BootloaderType.HalfKay;
        Name = "HalfKay";
        PreferredDriver = "HidUsb";
        IsResettable = true;
    }

    public override Task FlashAsync(string mcu, string file)
    {
        ValidateFileExtension(file, ".hex");
        return RunToolAsync("teensy_loader_cli", $"-mmcu={mcu}", file, "-v");
    }

    public override Task ResetAsync(string mcu) =>
        RunToolAsync("teensy_loader_cli", $"-mmcu={mcu}", "-bv");
}
